contracts/campaignAirdrops/ExistingERC20DropCampaign.sol

```diff
@@ -0,0 +1,126 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.14;
+
+import "@openzeppelin/contracts/token/ERC1155/ERC1155.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "../interfaces/AirdropMerkleProof.sol";
+import "../interfaces/IAirBroFactory.sol";
+
+/// @title Airdrops existing ERC20 tokens for airdrop recipients
+contract ExistingERC20DropCampaign is AirdropMerkleProof {
+    using SafeERC20 for IERC20;
+
+    IERC20 public immutable rewardToken;
+    address public immutable airbroCampaignFactoryAddress;
+    uint256 public immutable tokenSupply;
+
+    string public airdropType = "ERC20";
+    bool public airdropFunded;
+    bool public merkleRootSet;
+    uint256 internal numberOfClaimers;
+    uint256 public tokensPerClaim;
+    uint256 public airdropExpirationTimestamp;
+    bytes32 public merkleRoot;
+
+    mapping(address => bool) public hasClaimed;
+
+    address internal airdropFunder;
+
+    event MerkleRootChanged(bytes32 merkleRoot, uint256 _numberOfClaimers);
+    event Claimed(address indexed claimer);
+    event AirdropFunded(address contractAddress);
+
+    error AlreadyFunded();
+    error Unauthorized();
+    error AlreadyRedeemed();
+    error AirdropStillActive();
+    error AirdropExpired();
+    error NotEligible();
+    error InsufficientAmount();
+    error InsufficientLiquidity();
+    error MerkleRootAlreadySet();
+
+    modifier onlyAdmin() {
+        if (msg.sender != IAirBroFactory(airbroCampaignFactoryAddress).admin()) revert Unauthorized();
+        _;
+    }
+
+    constructor(
+        address _rewardToken,
+        uint256 _tokenSupply,
+        address _airbroCampaignFactoryAddress
+    ) {
+        rewardToken = IERC20(_rewardToken);
+        tokenSupply = _tokenSupply;
+        airbroCampaignFactoryAddress = _airbroCampaignFactoryAddress;
+    }
+
+    /// @notice Sets the merkleRoot and the number of claimers (also setting the amount each claimer receivers)
+    /// @notice can only be done by admin
+    /// @param _merkleRoot The root hash of the Merle Tree
+    /// @param _numberOfClaimers The number of people eligible to claim
+    function setMerkleRoot(bytes32 _merkleRoot, uint256 _numberOfClaimers) external onlyAdmin {
+        if (merkleRootSet) revert MerkleRootAlreadySet();
+
+        merkleRootSet = true;
+        merkleRoot = _merkleRoot;
+        airdropExpirationTimestamp = block.timestamp + 60 days;
+        numberOfClaimers = _numberOfClaimers;
+        tokensPerClaim = tokenSupply / numberOfClaimers;
+
+        emit MerkleRootChanged(_merkleRoot, _numberOfClaimers);
+    }
+
+    /// @notice Allows the airdrop creator to provide funds for the airdrop reward
+    function fundAirdrop() external {
+        if (airdropFunded) revert AlreadyFunded();
+
+        if (rewardToken.balanceOf(msg.sender) < tokenSupply) revert InsufficientAmount();
```

Comment:
`safeTransferFrom` on line 80 will revert if the msg.sender doesn't have the tokens / did not approve them.

Reply:
We perform checks for allowance and balance (line 79 and 80) prior to the safeTransfer on line 85, so this should be okay

Reply:
See comment for line 101

Reply:
safeTransfer already does the checks, variables set will revert if **safeTransfer** reverts